File: src/wfrp-helper-web/src/model/realm/Scenario.js
import {FormFieldType} from "../../form/FormFieldType";
import Person from "../creature/Person";
import Entity, {CreateColumn, CreateFormField, RegisterEntity} from "../Entity";
import Place from "../world/Place";
import Realm from "./Realm";

export default class Scenario extends Entity {
  static entityName: string = 'scenario';

  name: string;
  realm: Realm;
  places: Place[]   = [];
  persons: Person[] = [];

  updateWith(entity: Scenario): Scenario {
    this.updateProp(entity, 'id');
    this.updateProp(entity, 'uuid');
    this.updateProp(entity, 'name');
    this.updateEntityProp(entity, 'realm', () => new Realm());
    this.updateEntityList(entity, 'places', () => new Place());
    this.updateEntityList(entity, 'persons', () => new Person());
    return this;
  }
}

RegisterEntity(Scenario,
               [CreateColumn('name', FormFieldType.TEXT),
                CreateColumn('realm', FormFieldType.ENTITY_SELECT),],
               [CreateFormField('name', FormFieldType.TEXT),
                CreateFormField('realm', FormFieldType.ENTITY_SELECT, Realm.entityName),
                CreateFormField('places', FormFieldType.ENTITY_COMBOBOX, Place.entityName),
                CreateFormField('persons', FormFieldType.ENTITY_COMBOBOX, Person.entityName),]);

File: src/main/java/pl/khuzzuk/wfrp/helper/model/world/Scenario.java
package pl.khuzzuk.wfrp.helper.model.world;

import lombok.Getter;
import lombok.Setter;
import org.hibernate.validator.constraints.Length;
import pl.javahello.RemoteEntity;
import pl.javahello.RemoteEntity.SecuredService;
import pl.khuzzuk.wfrp.helper.model.creature.Person;
import pl.khuzzuk.wfrp.helper.repo.BaseEntity;

import javax.persistence.*;
import java.util.List;

@Getter
@Setter
@Entity
@Table(schema = "realm")
@RemoteEntity(transactional = true)
@SecuredService(allowRead = true)
public class Scenario extends BaseEntity {
  private @Length(min = 3, max = 255) String name;
  @ManyToOne
  private Realm realm;
  @ManyToMany
  @JoinTable(schema = "realm",
             name = "scenario_place",
             inverseJoinColumns = @JoinColumn(name = "place_id"))
  private List<Place> places;
  @ManyToMany
  @JoinTable(schema = "realm",
             name = "scenario_person",
             inverseJoinColumns = @JoinColumn(name = "person_id"))
  private List<Person> persons;
}
